Update events docs (#297)
paritytech/ink-docs: docs/basics/events.md

## Event Definition

> Since `ink!` version `5.0`, events can be defined independently of the contract which emits them. 

Reviewer: ```suggestion
Since ink! version 5.0, events can be defined independently of the contract which emits them. 
```

Events can now be defined once and shared across multiple contracts. 

> This is useful for events for contracts which conform to standards such as `ERC20`: 

Reviewer: ```suggestion
This is useful for events for contracts which conform to standards such as ERC-20: 
```

```rust
use ink::primitives::AccountId;

#[ink::event]
pub struct Transferred {
    #[ink(topic)]
    from: Option<AccountId>,
    #[ink(topic)]
    to: Option<AccountId>,
    amount: u128,
}
```
> Note that generics are [not currently supported](https://github.com/paritytech/ink/issues/2044)
> , so the concrete types of `Environment` 
> specific types such as `AccountId` must match up with the types used in the contract.

This definition can exist within a contract definition module (inline events), in a different 
module in the same crate or even in a different crate to be shared by multiple contracts.

### Legacy syntax for inline Event definitions

Events defined within a `#[ink::contract]` module can continue to use the original syntax for an 
event definition, using the `#[ink(event)]` attribute. Under the covers this is simply expanded 
to the new top level `#[ink::event]` macro, so both events defined using the legacy style and 
using the new `event` attribute macro directly will behave exactly the same.

### Topics

When an event is emitted, 0 or more topics can be associated with it. The event is then indexed 
together with other events with the same topic value.

> An event's fields can be annotated with `#[ink(topic)]` (see example), which will result in a 

Reviewer 2: ```suggestion
An event's field can be annotated with `#[ink(topic)]` (see example), which will result in a 
```

Author: It is meant to convey that multiple individual fields can be annotated with the `topic`.

topic derived from the value of that field being emitted together with the event.

> Topics are by default a 32 byte array (`[u8; 32]`), although this is configurable on the 

Reviewer 2: Is it actually configurable? I thought ink is hardcoded to use blake 256 still

Author: The type of the resulting topic is configurable yes. So if it is shorter I guess it will be truncated and longer padded.

https://github.com/paritytech/ink/blob/27407b5f59f24233cddc3cea59cc2213ba2c6674/crates/env/src/engine/on_chain/impls.rs#L157

Our way of generating the topic value itself is hardcoded to blake2.

above was not made indexable ‒ there will most probably be a lot of different events with differing
amounts each.

#### Signature Topic

By default all events have a signature topic. This allows indexing of all events of the same 
type, emitted by different contracts. The `#[ink::event]` macro generates a signature topic at 
compile time by hashing the name of the event concatenated with the *names of the types* of the all 
the field 
names:
```
blake2b("Event(field1_type,field2_type)")`
```
So for our `Transferred` example it will be: 
```
blake2b("Transferred(Option<AccountId>,Option<AccountId>,u128)")`
```
> Important caveat: because the *name* of the field type is used, refactoring an event 
> definition to use a type alias or a fully qualified type will change the signature topic, even
> though the underlying type is the same. Two otherwise identical definitions of an event with the 
> same name and same field types but different field type names will have different signature 
> topics.

When decoding events emitted from a contract, signature topics are now required to determine which 
type of event to decode into. 

#### Anonymous Events

> Annotating an event definitions with `#[ink(anonymous)]` (See [here](/macros-attributes/anonymous)

Reviewer: ```suggestion
Annotating an event definition with `#[ink(anonymous)]` (See [here](/macros-attributes/anonymous)
```
